_posts/2016/2016-11-21-multitenancy-and-elasticsearch.md

```diff
@@ -0,0 +1,127 @@
+You can mitigate that problem by running multiple installations on the same hardware, but every installation will have some fixed costs and this model is limited in terms of scalability.
+
+The multitenant architecture usually has higher fixed costs for a single installation, but the resources are shared and each tenant add considerably less overhead.
+
+There are also other benefits, you don't need to spin up a new installation to add a new tenant, you can usually add one almost instantaneously and maintenance is easier because you do it for all merchants simultaneously.
+
+This all sounds great, but multitenancy comes with its own challenges.
+
+### Why not?
+
+A multitenancy architecture is not the cure-all for every application and scaling problem out there and there are some general drawbacks.
+
+The major one is data isolation, you don't want to accidentally leak or mix sensitive information of different tenants.
+
+Another downside is the complexity you introduce - the complexity of the system itself as well as the complexity the developer has to deal when implementing features in a tenant context.
+This depends on the frameworks you use and abstractions you introduce.
+
+## Elasticsearch
+
+As already mentioned, there are some challenges when it comes to multitenancy, especially when it comes to data isolation and reducing the development complexity overhead.
+
+Since multitenancy is not a novel concept, a lot of frameworks and storage-solutions have some support for multitenancy out of the box.
+Elasticsearch and the frameworks we evaluated do not come with any such a feature and in addition, the way to implement that is not as straight forward as it is for traditional databases.
+
+There are two general ways to go about it, which are
+
+1. One index per tenant
+2. Shared index
+
+### One Index per Tenant
+
+The first option is an easy and straightforward concept.
+It provides solid data isolation from a storage perspective and you can still query across multiple indexes if you need to. This solution is sufficient for most cases.
+
+Each index comes with some overhead and this can easily add up.
+By introducing a new index per tenant, you are multiplying your index by the number of tenants `T` you have.
+If you then introduce localization for `L` languages and need to create a new index for that as well.
+In each of those, you might have `T` different types, which you might (or cannot) keep in the same index as well.
+Considering that, you end up with `T * L * T` indices for one installation.
+As an example, for 10,000 tenants, 10 types and 10 supported languages, you will need 1 million indices.
+
+### Shared Index
+
+To keep the amount of indices under control, you can share the same index for all tenants.
+This would reduce the number of indices to 100,000 for the example mentioned above.
+
+Elasticsearch provides some features that help in dealing with multitenancy in a single index.
```

Comment:
Elasticsearch provides some features that help in dealing with multitenancy in a single index. >
Elasticsearch provides some features that help dealing with multitenancy in a single index.